### ISSUE TITLE
MLE PCA Selects Large Numbers of Components

### ISSUE THREAD

**Issue by jbteves** [ME-ICA/tedana#356]
### Summary
In the `mle` option, the number of surviving PCA components is always the number of TRs minus 1, which is a large number of components for fMRI data.

### Additional Detail
We're currently unsure of why there are so many components in the fMRI data, but it seems to be massively over-fitted. 

@handwerkerd thinks that the lack of variance-normalization may be a factor.

### Next Steps
- [ ] Discuss why the number of components appears to be so large
- [ ] Provide a fix that reduces the number of components to something more reasonable.

**Comment by jbteves** [comment]
Note: I've posted about this on [neurostars](https://neurostars.org/t/number-of-components-trs-1-in-tedana/3339) and another poster also indicated the problem. @handwerkerd and @smoia have also indicated that this behavior deviates from what one should expect.

**Comment by smoia** [comment]
Running Melodic on the same optimally combined dataset that serves as input for the PCA step in tedana, gives a different dimensionality estimation. On a dataset that I'm using, the number of components returned by Melodic is 76 against 319 estimated by tedana-mle (which intriguingly is also the number of components estimated by tedana-kundu - not stabilised).

This comparison is interesting to me for at least two reasons.
First, Melodic is accepted as a valid option to run ICA on fMRI data, and is the base for denoising options (AROMA and FIX), that are integrated in community tools as well as potentially integrated in post-tedana workflows.
Second, reading Melodic's source code (it's in C++, fun reading if you want to go for it), as far as decomposition goes, sklearn's PCA run with the MLE option should be exactly the same as the PCA incorporated in Melodic. FastICA should be exactly the same ICA run by Melodic, too.

I believe the number of components estimated by tedana-mle and Melodic shouldn't be exactly the same, but this huge difference in number is quite suspicious to me, due to the fact that the two software should be doing very similar steps.

Hence, in order to understand and possibly find a better dimensionality estimator, I would try to understand the differences between Melodic and tedana-mle.

If I got the code right, in tedana:
1) Data are variance-normalised before performing PCA
2) Are AR models taken into consideration in order to have residuals that are not autocorrelated?
3) Melodic performs a Cholewsky decomopsition in order to retrieves a lower triangular matrix from the noise covariance matrix, and its inverse is used to make the data uncorrelated from the noise. This step is not present anywhere in tedana (nor in the libraries used by tedana). Am I right? I'm not saying it should, but maybe that's a source of MLE estimations being so big.
4) Noise is taken into account in the PCA model, being modelled as a term of the equation. Isn't it?
5) Another  difference is that in Melodic, the eigenspectrum of the data is adjusted, comparing it against the (expected) eigenspectrum of random Gaussian matrix. This step might be part of PCA in sklearm, but I didn't find it.

**Comment by tsalo** [comment]
Just want to note that #218 exists (it's closed, but it's there), in case you want to steer more MELODIC-specific discussions there.

Regarding MLE, has anyone tested `--sourceTEs` to see if using the multi-echo data for PCA instead of the optimally combined data would improve the dimensionality estimation?

> 2. Are AR models taken into consideration in order to have residuals that are not autocorrelated?

Unless that's something that's incorporated into the standard PCA method (which I can't see why it would be, given that most data fed into PCA shouldn't be autocorrelated), I'm pretty sure this is a no.

> 5. Another difference is that in Melodic, the eigenspectrum of the data is adjusted, comparing it against the (expected) eigenspectrum of random Gaussian matrix. This step might be part of PCA in sklearn, but I didn't find it.

That sounds like the "dimensionality estimation" step. It sort of sounds like the MP-PCA approach, but maybe it's similar to MLE?

**Comment by handwerkerd** [comment]
I don't think this is about whether or not to specifically use the melodic approach. The current issues is independent of whether one uses the separate echos or the optimally combined volumes. Dimensionality reduction should significantly reduce the number of dimensions in fMRI data and it isn't. I suspect @smoia is looking in the right direction and the current code is missing a critical data normalization step.

**Comment by emdupre** [comment]
In gitter @smoia asked about accessing the MEICA 2.5 dimensionality reduction -- yes, you can in tedana ! It's just calling `kundu` with the `--tedpca` argument.

**Comment by smoia** [comment]
Thanks @emdupre, I thought so but wanted to be sure. Funny thing - the dimensionality estimation with tedpca decision tree is the same as MLE. I didn't try using the code of meica2.5.

**Comment by jbteves** [comment]
Per the call, there seems to be strong feelings that the issue stems from a bug in normalizing the variance explained, especially since the same thing happens in both `kundu` and `mle` cases.

**Comment by smoia** [comment]
I'm not sure that the issue with the `kundu` case is related to normalising variance or an issue with the elbow selection. For the `mle` case, that should be the case.

**Comment by emdupre** [comment]
Here's a very simple proposed fix -- what if we just request a specific amount of the variance instead of the MLE option, setting the threshold ourselves ? From [the sklearn documentation](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html):

> If 0 < n_components < 1 and svd_solver == 'full', select the number of components such that the amount of variance that needs to be explained is greater than the percentage specified by n_components.

The question would be what to set the percentage as. I'd suggest 90% as a starting point. We could run a few tests to see how this impacts, but does anyone have specific benchmarks from the literature to recommend ?

**Comment by handwerkerd** [comment]
This might be better than the status quo, but it also risks making something look correct when it isn't. The appropriate percent of variance retained will heavily depend on the inputted data. For example, if you use smaller voxels, with  lower SNR, then one would expect more thermal noise and less variance explained by the ICA.

That said, if we set the threshold at the upper bar (i.e. Has anyone ever seen more than 96% variance explained for fMRI data?) then it will always create too many components, but still give us much fewer than we're currently getting.

I'm still pushing that there is a relatively minor code fix for this and the right person just has to look at the code to figure out what we're doing wrong. Next week, I'll try to look at the code with @javiergcas to see if he can figure it out.

**Comment by emdupre** [comment]
It's just the PCA variance explained, here, so the ICA should correctly then narrow down to the correct overall variance. All that to say, I think this idea of a reasonable upper bar is exactly what the proposed strategy could provide.

But I'd be happy to see the PPCA approach with MLE work ! I'm also optimistic it's something small, but I haven't been able to look in depth at what it could be. Looking forward to what you and @javiergcas find !

**Comment by tsalo** [comment]
Given that we will add new PCA options in #435 and MLE is still broken, is the current plan to just drop MLE as an option?

**Comment by emdupre** [comment]
>  is the current plan to just drop MLE as an option?

Yes, I think so. I think we should also implement a threshold based approach as I discussed above (https://github.com/ME-ICA/tedana/issues/356#issuecomment-528954929). I mentioned this to Gaël and he suggested using 95% of the variance. Are there strong opinions for another threshold ?

**Comment by tsalo** [comment]
> I think we should also implement a threshold based approach as I discussed above (#356 (comment)).

That's great. Would this be for `kundu` and `kundu-stabilize` specifically? I think that would address my concerns regarding the z-statistic calculation within PCA.

> I mentioned this to Gaël and he suggested using 95% of the variance. Are there strong opinions for another threshold?

I'm happy with 95%. It's probably worth it to determine how many components that tends to remove, to see how _that_ would impact the validity of our z-statistics.

**Comment by emdupre** [comment]
> That's great. Would this be for kundu and kundu-stabilize specifically? I think that would address my concerns regarding the z-statistic calculation within PCA.

No, sorry, this would be another option ! `kundu` is still the original method of applying a decision tree to the PCA components to do the selection, and `kundu-stabilize` applies the 95% variance threshold on top of the decision tree (see [here](https://github.com/ME-ICA/tedana/blob/5e0b3eefddfe751a0964b048bb4389d066028e28/tedana/selection/tedpca.py#L109)).

I'm suggesting we have a method just called something like `pca` where it's a 95% variance threshold but no decision tree !

**Comment by tsalo** [comment]
> kundu is still the original method of applying a decision tree to the PCA components to do the selection

Since the initial decision tree in the PCA will generate `N volumes` components, metrics relying on conventional z-statistic calculation will be invalid, so I'm not a fan of keeping `kundu` the way it is.

> I'm suggesting we have a method just called something like `pca` where it's a 95% variance threshold but no decision tree !

I see. I think that's a reasonable option, although in that case it's even more important to see how many components tend to be removed with that threshold. I was assuming that the 95% applied to the decision-tree-based PCA methods, in which case having too few degrees of freedom would only adversely impact the PCA metrics and the ICA metrics would probably be fine (since the decision tree would hopefully remove enough additional components to have sufficient degrees of freedom). If the PCA returns the top 95% without further reduction, then the number of components will impact the ICA metrics.

**Comment by smoia** [comment]
@tsalo , I think that the patch @CesarCaballeroGaudes is working on regarding a different application of z statistic can be applied in any possible case. It actually changes the results quite a lot, and will be valid despite differences in `N volumes - 1` components.

**Comment by handwerkerd** [comment]
I think I'm missing something here. My understanding is that #435 is a principled way of defining the signal that will be input to the ICA. Setting a fixed percentile can be a good approximation but, since signal vs gaussian noise varies across FMRI acquisitions, a fixed valued will not be consistently correct. I could see adding an option "Fixed variance PCA" which would be useful for testing and possibly other situations, but that shouldn't be presented as an appropriate option.

If there's going to be a long time lab before #435 is merged, I can see having a fixed variance as a good temporary option, but it should be treated as such.

As for the kundu option, I think the key innovation is to examine low variance components that otherwise would have been rejected and retain them if they have high kappa or rho values. This innovation could be added to any PCA method rather than being treated as a separate option. I could see having such a "keep high kappa or rho PCA components" option that could be used with maPCA or fixed variance PCA.

**Comment by CesarCaballeroGaudes** [comment]
maPCA will return a  lower number of components than MLE and probably all of them with high variance. Therefore, maPCA will likely not be appropriate in combination with kundu or kundu-stabilize options if the aim is to retain components with low variance and high kappa or rho values. Of course, both kundu options, which are performed after PCA order estimation, can then be used to further diminish the number of components.

**Comment by handwerkerd** [comment]
I'm still a bit confused here. First, my understanding is maPCA is how MLE is implemented in Calhoun's GIFT ICA. @CesarCaballeroGaudes by comparing the two are you saying maPCA is completely distinct from MLE?

On the main question here, my understanding of the kundu approach is that the idea is to select components based on kappa or rho weighting instead of variance. The ideal is the long-term goal of modeling echo dependence directly in the decomposition step. The intermediate idea is to use some other method to select components based on variance. For example, maPCA will define which components contain signal and which contain noise and it would only pass the signal components to ICA. With an add on, after maPCA decides X components are noise, check if any of those components have a non-trivial fit to a kappa or rho model. If they fit a model, then they are inherently not noise and shouldn't be discarded. Thus this is a way to retain a small about of additional variance that would have otherwise been rejected by the PCA dimensionality reduction step. I don't know if this matters or not and it should be applied with caution, but I don't think it's an inherently flawed approach.

More generally, whether or not it's a good idea, I don't think the kundu approach is linked to any specific dimensionality reduct step so it probably shouldn't be coded as it's own dimensionality reduction step.

**Comment by CesarCaballeroGaudes** [comment]
MLE and maPCA are two algorithms to estimate the number of components in PCA decomposition, which is usually used for ICA decomposition too. MLE approach is based on Minka's 2001 paper, whereas the maPCA code is a copy of the algorithm used in GIFT ICA (described in https://doi.org/10.1002/hbm.20359). Both algorithms have the same purpose, but have different foundations and thus yield different order estimates.
The issue here is that among the options for estimating the order [in pca.py](https://github.com/ME-ICA/tedana/blob/98e8590e5fbf06b5541f0112c4ed7020178e5ea2/tedana/decomposition/pca.py#L255-L267), the order for algorithm="mle" is in practice NT-1, which is equal to the order that would be used for "kundu and "kundu-stabilize" (i.e. the 'else' option in the if-statement). Therefore, at this point of the code, the order will always be NT-1 regardless of algorithm, either "mle", "kundu" or "kundu-stabilize".  The maPCA approaches yields a lower order estimate.
Next in the code, the options 'kundu' and 'kundu-stabilize' will classify principal components based on rho, kappa and variance [afterwards](https://github.com/ME-ICA/tedana/blob/98e8590e5fbf06b5541f0112c4ed7020178e5ea2/tedana/decomposition/pca.py#L296-L304). If the aim is to retain components with low variance, which will likely be part of the noise part in maPCA, then the 'kundu' or 'kundu-stabilize' decisions must be applied on all of the NT-1 components and the current approach is fine. However, we are seeing that the number of components is still very large with  'kundu' or 'kundu-stabilize'.
Alternatively, the maPCA could be used to estimate a lower order that are always accepted, and then pass some of the excluded components to "kundu" and "kund-stabilize"  to further classify and accept a subset of them before [performing the PCA projection.](https://github.com/ME-ICA/tedana/blob/98e8590e5fbf06b5541f0112c4ed7020178e5ea2/tedana/decomposition/pca.py#L324-L328). Of course the number of retained components in this step should be much lower than right now. Otherwise, the maPCA will not serve its purpose. As a suggestion, these options can be "maPCA-kundu" and "maPCA-kundu-stabilize".

**Comment by smoia** [comment]
I hope not to throw more confusion with this, but:
My understanding of the original purpose of the `kundu` option (as of `meica` v2, pre-`tedana`) was to implement [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3791700/).
In the [supplementary materials](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3791700/bin/1301725110_pnas.201301725SI.pdf) there's a section titled "TE Dependence–Based Dimensionality Reduction", in which it's explained that the concept behind `tedpca` is to apply a dimensionality reduction after a full SVD decomposition, in which the criteria selection is based on rho, kappa and variance scores above a threshold (threshold found using the elbow method).
If this is the decision tree you're talking about, then I don't necessarily understand where z-statistics play a role - this should be done on raw scores coming from fitting rho and kappa to the timeseries of the components. Otherwise, I'm with @tsalo in rethinking the way `kundu` is run.

In both cases, `kundu` becomes a full alternative to `mle` or `maPCA`, not something to be applied on top of those dimensionality selection criteria.
The paper states that the threshold on variance is virtually redundant because high kappa or high rho corresponds to high variance. This can be read as a) we shouldn't expect any case with low variance and high rho/kappa, b) running `maPCA` would be a shortcut for a simpler resolution of `kundu`. However, there's no observation that this correspondence is actually real - so I wouldn't opt for running `kundu` on top of `maPCA` as part of the signal to categorise would be lost.
